### PR TITLE
Step retries, Scenario result and skipped steps.

### PR DESCRIPTION
**Pull request by AFASbart**
When a step fails, it will retry the same step for a configurable time (default is 3). If any of the steps actually fails, it will be marked as failed and the other steps will be marked as skipped (it will not be run but all the hooks etc will be run).

When a step has failed, all the hooks will be reminded that it has failed / passed and the JSON reported will have the scenario as passed: true or passed: false.

Fixed issues: #218, #201 and #179.

### CODE REVIEW

**Comment by AFASbart** [comment]
@jonsamwell I also have created a pull request on the dart_gherkin package. I require that package to be updated and this package to have that newer version before we can release a new rc.

Can you create new releases for me?

File: lib/src/flutter/runners/gherkin_integration_test_runner.dart [hunk 1]
```diff
@@ -254,12 +271,8 @@ abstract class GherkinIntegrationTestRunner {
...
 
   @protected
-  Future<StepResult> runStep(
-    String step,
-    Iterable<String> multiLineStrings,
-    dynamic table,
-    TestDependencies dependencies,
-  ) async {
+  Future<StepResult> runStep(String step, Iterable<String> multiLineStrings,
+      dynamic table, TestDependencies dependencies, bool hasToSkip) async {
```
Comment (jonsamwell) [inline lib/src/flutter/runners/gherkin_integration_test_runner.dart:275]: Please add a comma after `hasToSkip` we should probably get the latests flutter_lints packages sorted in library

File: lib/src/flutter/runners/gherkin_integration_test_runner.dart [hunk 2]
```diff
@@ -135,7 +137,8 @@ abstract class GherkinIntegrationTestRunner {
   void runScenario(
     String name,
     Iterable<String>? tags,
-    Future<void> Function(TestDependencies dependencies) runTest, {
+    List<Future<StepResult> Function(TestDependencies dependencies, bool skip)>
```
Comment (jonsamwell) [inline lib/src/flutter/runners/gherkin_integration_test_runner.dart:140]: Add trailing comma

File: lib/src/flutter/runners/gherkin_integration_test_runner.dart [hunk 3]
```diff
@@ -283,28 +296,43 @@ abstract class GherkinIntegrationTestRunner {
...
-      configuration.defaultTimeout,
-      parameters,
-    );
+    StepResult? result;
 
+    if (hasToSkip) {
+      result = new StepResult(
+          0, StepExecutionResult.skipped, "Previous step(s) failed.");
+    } else {
+      for (int i = 0; i < this.configuration.stepMaxRetries; i++) {
```
Comment (jonsamwell) [inline lib/src/flutter/runners/gherkin_integration_test_runner.dart:305]: When did `stepMaxRetries` get added to the configuration? Was this in a previous PR?

**Review by jonsamwell — CHANGES_REQUESTED** [review]
Looks good, just a few lint fixed :-)

File: lib/src/flutter/runners/gherkin_integration_test_runner.dart (hunk 3 above)
Comment (AFASbart) [reply lib/src/flutter/runners/gherkin_integration_test_runner.dart:305]: This has now been added in the dart_gherkin package. I have also created a PR there. That is why it is required for dart_gherkin to be updated before this package.

Comment (jonsamwell) [reply lib/src/flutter/runners/gherkin_integration_test_runner.dart:305]: Make sense!

**Comment by jonsamwell** [comment]
Nice work! Thanks

**Comment by umangdivyanshu** [comment]
Thanks @AFASbart for the great work. 💯 

@jonsamwell Would there be a new release for this?

**Comment by jonsamwell** [comment]
@umangdivyanshu I think we are just waiting on merging another pr then we can release - @AFASbart is this correct?

**Comment by AFASbart** [comment]
Yes I have added a comment to your comment on PR #221, can you take a look? @jonsamwell

**Comment by anormaly-dev** [comment]
Hello @AFASbart , @jonsamwell , I am not sure if is the correct place to ask this... but I don't understand how to enable the retry mode. Now when I create `FlutterTestConfiguration` I don't have any option to set the `stepMaxRetries`, and by default in `TestConfiguration` is `0`.

**Comment by AFASbart** [comment]
If I look into the code, I see that the field is final. That means that you have to set the field in the constructor / during initialization. You can look up online how to set final fields in dart. @anormaly-dev

**Pull request merged**